src/lib/state-machine/state-machine.ts

```diff
@@ -0,0 +1,48 @@
+import { StateMachine, MachineEvent, MachineState, MachineNode, MachineTransitionFn } from '../state-machine-api';
+
+export class GenericStateMachine<TOutput> implements StateMachine<TOutput> {
+  private currentState: MachineState<TOutput>;
+  private closed = false;
+
+  constructor(
+    private initialNode: MachineNode,
+    private finalNodes: Set<MachineNode>,
+    private transitionFn: MachineTransitionFn<TOutput>,
+  ) {
+    this.currentState = {
+      currentNode: initialNode,
+    };
+  }
+
+  public getCurrentState(): MachineState<TOutput> {
+    return this.currentState;
+  }
+
+  public handle($event: MachineEvent): MachineState<TOutput> | never {
+    if (this.isClosed()) {
+      throw new Error('StateMachine is now closed and can not accept any more events');
+    }
+    this.currentState = this.transitionFn(this.currentState, $event);
+    return this.currentState;
+  }
+
+  public canClose(): boolean {
+    return !this.isClosed() && this.finalNodes.has(this.currentState.currentNode);
```

Comment:
```suggestion
    return !this.isClosed() && this.finalNodes.has(this.currentState?.currentNode);
```

Reply:
No nullchecks, and that is a fancy nullcheck.
CurrentState must never be null/undefined.